enhancements/installer/vsphere-ipi-multiple-datacenter-clusters.md

```diff
@@ -0,0 +1,459 @@
+---
+title: multiple-datacenters-cluster-vsphere-ipi
+authors:
+  - "@jcpowermac"
+reviewers:
+  - "@rvanderp3"
+  - "@bostrt"
+  - ""
+approvers:
+  - ""
+creation-date: 2021-09-21
+last-updated: 2021-09-21
+status: implementable
+see-also:
+  - "/enhancements/"
+replaces:
+  -
+superseded-by:
+  -
+---
+
+# Multiple vCenter datacenters and clusters in vSphere IPI
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The goal of this enhancement is to provide the ability to install in a
+vSphere enviornment with multiple datacenters and clusters. Including configuration of regions and zones in the CCM.
+
+## Motivation
+
+Users of OpenShift would like the ability to deploy
+within multiple datacenters and clusters to increase
+reliablity. Customers would also like to take advantage
+of the concept of regions and zoning that this type of deployment would
+offer.
+
+### Goals
+
+- Support regions and zones in vSphere using multiple datacenters (region) and clusters (zone)
+- Support installation into multiple datacenters and multiple clusters
+
+### Non-Goals
+
+- Support multiple subnets
+- Support multiple vcenters
+
+## Proposal
+
+Modification of the installer to support the provisioning of masters in defined datacenters and clusters.
+
+### installer
+
+#### Platform spec
+
+Keep the vSphere platform spec unchanged.
+
+#### MachinePool
+
+The `MachinePool` will need to be significantly modified to support this effort.
+
+```golang
+type MachinePool struct {
+	NumCPUs int32 `json:"cpus"`
+	NumCoresPerSocket int32 `json:"coresPerSocket"`
+	MemoryMiB int64 `json:"memoryMB"`
+	OSDisk `json:"osDisk"`
+	VCenter []`json:"vcenter,omitempty"`
```

Comment:
Would it simplify things to bring the vCenter specifications to the top-level platform rather than the machine pool?
1) Move vCenter specifications to `.platform.vsphere`.
2) Have the MachinePool reference which vCenters to use by name specified in `.platform.vsphere`.

This would bring it more in line with how regions and zones are used in cloud platforms. It would also avoid duplication in the case where control plane and computes nodes are in the same data centers and clusters (which I am guessing will be the most common case).

Reply:
Thanks @staebler for taking a look. Were you thinking something like this?

```yaml
apiVersion: v1
baseDomain: example.com
controlPlane:
  name: master
  platform:
    vsphere:
      zones:
      - us-west-1a
      - us-east-1b
      cpus: 8
      ...
  replicas: 3
compute:
- name: worker
  platform:
    vsphere:
      zones:
      - us-west-1a
      - us-east-1b
      cpus: 8
      ...
  replicas: 3
metadata:
  name: test-cluster
platform:
  vSphere:
    vCenter: your.vcenter.example.com
    ...
    locations:
      regions:
      - name: us-east
        zones:
        - name: us-east-1a
          datacenter: dc1
          cluster: dc1_c1
          resourcepool: /DC1/host/DC1_C1/Resources
          network: "vm network"
          datastore: "DC1_DS1"
          folder: "vmfolder"
        - name: us-east-1b
          datacenter: dc1
          cluster: dc1_c2
      - name: us-west
        zones:
        - name: us-west-1a
          datacenter: dc2
          cluster: dc2_c1
        - name: us-west-1b
          datacenter: dc2
          cluster: dc2_c2
```
I started to think if we have to extend this to support multple vcenter(s) that the eventual ability to specific a vcenter configuration in the regions object might be a positive.

Reply:
Thanks again @staebler for bouncing ideas...

Taking a look at:
https://cloud-provider-vsphere.sigs.k8s.io/tutorials/deploying_cpi_and_csi_with_multi_dc_vc_aka_zones.html

I think we should keep region and move the datacenter parameter to it - like you mentioned above:
https://github.com/openshift/enhancements/pull/918#discussion_r731948025

If we do that we can create the tag and tag category in terraform instead of that being required for the customer to do day0. 
Then force the opinionated design: a datacenter is a region, a cluster is a zone.
wdyt?